Integration tests: transferTx_value, transferTx_from

churik opened the pull request:
Added first parametrized tests from checking errors in create_multi_transaction method (note, that I just put a random `str` to `from` to show that it is easily expandable)
In future will expand with more values

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (18)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 28cf171b | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/1/) | 2024-09-05 12:17:26 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/1//consoleText) |
| :heavy_check_mark: | 28cf171b | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/1/) | 2024-09-05 12:18:52 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/1//consoleText) |
| :heavy_check_mark: | 28cf171b | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5802/1/) | 2024-09-05 12:20:31 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240905-121605-28cf17-pr5802.zip) |
| :heavy_check_mark: | 28cf171b | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5802/1/) | 2024-09-05 12:21:14 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240905-121604-28cf17-pr5802.zip) |
| :heavy_check_mark: | 28cf171b | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5802/1/) | 2024-09-05 12:21:54 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240905-121604-28cf17-pr5802.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 0d034176 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/2/) | 2024-09-09 16:44:22 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/2//consoleText) |
| :heavy_check_mark: | 0d034176 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5802/2/) | 2024-09-09 16:44:47 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240909-164305-0d0341-pr5802.aar) |
| :heavy_check_mark: | 0d034176 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5802/2/) | 2024-09-09 16:45:19 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240909-164307-0d0341-pr5802.zip) |
| :heavy_multiplication_x: | 0d034176 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/2/) | 2024-09-09 16:45:21 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/2//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | 14b82224 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/3/) | 2024-09-09 16:50:25 | ~53 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/3//consoleText) |
| :heavy_check_mark: | 14b82224 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5802/3/) | 2024-09-09 16:51:04 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240909-164925-14b822-pr5802.aar) |
| :heavy_check_mark: | 14b82224 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5802/3/) | 2024-09-09 16:51:34 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240909-164927-14b822-pr5802.zip) |
| :heavy_multiplication_x: | 14b82224 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/3/) | 2024-09-09 16:51:42 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/3//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | cc2e9ef2 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/4/) | 2024-09-18 16:02:44 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/4//consoleText) |
| :heavy_multiplication_x: | cc2e9ef2 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/4/) | 2024-09-18 16:03:46 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/4//consoleText) |
| :heavy_check_mark: | cc2e9ef2 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5802/4/) | 2024-09-18 16:05:09 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240918-160122-cc2e9e-pr5802.zip) |
| :heavy_check_mark: | cc2e9ef2 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5802/4/) | 2024-09-18 16:05:54 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240918-160117-cc2e9e-pr5802.aar) |
| :heavy_check_mark: | cc2e9ef2 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5802/4/) | 2024-09-18 16:06:02 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240918-160118-cc2e9e-pr5802.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | b3bfa4f3 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/5/) | 2024-09-18 16:50:16 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/5//consoleText) |
| :heavy_check_mark: | b3bfa4f3 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5802/5/) | 2024-09-18 16:50:33 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240918-164902-b3bfa4-pr5802.aar) |
| :heavy_check_mark: | b3bfa4f3 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5802/5/) | 2024-09-18 16:50:58 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240918-164906-b3bfa4-pr5802.zip) |
| :heavy_check_mark: | b3bfa4f3 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/5/) | 2024-09-18 16:51:23 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/5//consoleText) |
| :heavy_check_mark: | b3bfa4f3 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5802/5/) | 2024-09-18 16:52:06 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240918-164903-b3bfa4-pr5802.zip) |
| | | | | | | |
| :heavy_check_mark: | 3a7a6239 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5802/6/) | 2024-09-18 16:56:03 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240918-165423-3a7a62-pr5802.aar) |
| :heavy_check_mark: | 3a7a6239 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5802/6/) | 2024-09-18 16:56:31 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240918-165427-3a7a62-pr5802.zip) |
| :heavy_check_mark: | 3a7a6239 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/6/) | 2024-09-18 16:56:43 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5802/6//consoleText) |
| :heavy_check_mark: | 3a7a6239 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5802/6/) | 2024-09-18 16:57:50 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240918-165424-3a7a62-pr5802.zip) |
| :heavy_check_mark: | 3a7a6239 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/6/) | 2024-09-18 17:26:30 | ~31 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5802/6//consoleText) |

antdanchenko commented on integration-tests/tests/test_cases.py:
```diff
@@ -9,16 +9,16 @@ class RpcTestCase:
         self.network_id = 31337
 
-    def verify_is_valid_json_rpc_response(self, response, _id=None):
+    def verify_is_valid_json_rpc_response(self, response, _id=None, _error=False):
```
let's create a separate method, verify_is_jscon_rpc_error

antdanchenko commented on integration-tests/tests/test_cases.py:
```diff
@@ -29,6 +29,19 @@ def verify_is_valid_json_rpc_response(self, response, _id=None):
                 raise AssertionError(f"no id in response {response.json()}")
         return response
 
+    def validate_error_code_content(self, response, error_code: int, error_content):
```
I would appreciate to keep it as assert error code and assert error content in the test itself, meanwhile https://github.com/status-im/status-go/pull/5802/files#r1745360470 will be called first to ensure that we have `error` instead of result in response

antdanchenko commented on integration-tests/tests/test_cases.py:
```diff
@@ -46,9 +59,18 @@ def verify_json_schema(self, response, method):
 
-    def wallet_create_multi_transaction(self):
+    @staticmethod
+    def _update_multitx_method(params, updates):
```
it is advanced, on my opinion for better readability would be great to pass each parameter explicitly to wallet_create_multi_transaction

churik commented:
@antdanchenko 
review please one more time, thanks!

antdanchenko commented on integration-tests/tests/test_wallet_rpc.py:
```diff
@@ -39,14 +39,45 @@ def test_tx_(self, method, params):
+            raise AssertionError(
+                f"no required text in message: {actual_error_text} instead of expected: {expected_error_text}"
+            )
+        with open(f"{option.base_dir}/schemas/wallet_createMultiTransaction/transferTx_error", "r") as schema:
```
sorry, didn't notice on first review, we can use https://github.com/status-im/status-go/blob/develop/integration-tests/tests/test_cases.py#L45 here, I will also update other tests

churik replied:
can you elaborate please?

antdanchenko replied:
I meant to use `verify_json_schema` method

antdanchenko commented on integration-tests/tests/test_wallet_rpc.py:
```diff
@@ -39,14 +39,45 @@ def test_tx_(self, method, params):
+        response = self.wallet_create_multi_transaction(**changed_values)
+        self.verify_is_json_rpc_error(response)
+        actual_error_code, actual_error_text = response.json()['error']['code'], response.json()['error']['message']
+        if expected_error_code != actual_error_code:
```
I would prefer to have native asserts, e.g.
`assert  actual_error_code != expected_error_code, f"got code: {actual_error_code} instead of expected: {expected_error_code}"`

as result error message will be 
```
AssertionError: got code: -32000 instead of expected: 20000
```

also saves some space :)

churik replied:
done

antdanchenko commented on integration-tests/tests/test_wallet_rpc.py:
```diff
@@ -39,14 +39,45 @@ def test_tx_(self, method, params):
+            raise AssertionError(
+                f"got code: {actual_error_code} instead of expected: {expected_error_code}"
+        )
+        if expected_error_text not in actual_error_text:
```
we can do same as https://github.com/status-im/status-go/pull/5802/files#r1756993745 but with assert not in

churik replied:
done

churik commented:
~@antdanchenko I addressed comments, although I'm not sure why Jenkins can't find the files and why it is trying to find it in another dir, `/tests-rpc/`:~
```
| =================================== FAILURES ===================================

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | _______________ TestTransactionRpc.test_create_multi_transaction _______________

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | tests/test_wallet_rpc.py:47: in test_create_multi_transaction

[2024-09-18T16:03:44.121Z] tests-rpc_1                     |     with open(f"{option.base_dir}/schemas/wallet_createMultiTransaction/transferTx_positive", "r") as schema:

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | E   FileNotFoundError: [Errno 2] No such file or directory: '/tests-rpc/schemas/wallet_createMultiTransaction/transferTx_positive'

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | _ TestTransactionRpc.test_create_multi_transaction_validation[transferTx_value_not_enough_balance-changed_values0--32000-Insufficient funds for gas] _

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | tests/test_wallet_rpc.py:72: in test_create_multi_transaction_validation

[2024-09-18T16:03:44.121Z] tests-rpc_1                     |     with open(f"{option.base_dir}/schemas/wallet_createMultiTransaction/transferTx_error", "r") as schema:

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | E   FileNotFoundError: [Errno 2] No such file or directory: '/tests-rpc/schemas/wallet_createMultiTransaction/transferTx_error'

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | _ TestTransactionRpc.test_create_multi_transaction_validation[transferTx_from_from_invalid_string-changed_values1--32602-cannot unmarshal hex string without 0x prefix] _

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | tests/test_wallet_rpc.py:72: in test_create_multi_transaction_validation

[2024-09-18T16:03:44.121Z] tests-rpc_1                     |     with open(f"{option.base_dir}/schemas/wallet_createMultiTransaction/transferTx_error", "r") as schema:

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | E   FileNotFoundError: [Errno 2] No such file or directory: '/tests-rpc/schemas/wallet_createMultiTransaction/transferTx_error'

[2024-09-18T16:03:44.121Z] tests-rpc_1                     | =============================== warnings summary ===============================
```

as they are committed to this branch.

Should I just store them in main dir?


**UPDATE: turned out (who would have thought?) that Linux is case-sensitive for paths and Mac is not. 
Small step for humanity, big step for me :)**

churik committed:
test_: added negative tests for MutiTx

saledjenic reviewed (approved):
Looks ok to me, just want to share that `wallet_createMultiTransaction` is deprecated, probably will be removed at some point after mobile app switches to the new sending tx flow.

churik commented:
We'll re-do it anyway but makes sense to align with mobile client
